draft/0005-impersonation.md

```diff
@@ -0,0 +1,279 @@
+
+```
+POST /request-for-subject-token HTTP/1.1
+Host: resource.example.com
+Authorization: Bearer ...
+Content-Type: application/json
+
+{
+  "userId":"123456",
+  "ticketId":"7890"
+}
+```
+
+The resource owner API validates the support engineer's request and call Logto's management API using a machine-to-machine application to issue a subject token for the user.
+
+```
+POST /api/security/subject-tokens HTTP/1.1
+Host: example.logto.app
+Authorization: Bearer eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9...
+Content-Type: application/json
+
+{
+  "userId":"123456",
+  "expiresIn":600
+}
+```
+
+Logto validates the request and issues a subject token for the user.
+
+```
+HTTP/1.1 200 OK
+Content-Type: application/json
+
+{
+  "subjectToken":"accVkjcJyb4BWCxGsndESCJQbdFMogUC5PbRDqceLTC",
+  "expiresIn":600
+}
+```
+
+The resource owner API returns the subject token to the support engineer.
+
+### Step 2: Exchange the subject token for an access token
+
+The support engineer exchanges the subject token for an access token that represents the identity of the user. The support engineer can use the access token to make API requests on behalf of the user.
+
+```
+POST /oidc/token HTTP/1.1
+Host: example.logto.app
+Content-Type: application/x-www-form-urlencoded
+
+grant_type=urn%3Aietf%3Aparams%3Aoauth%3Agrant-type%3Atoken-exchange
+&scope=openid%20offline_access%20profile
+&subject_token=accVkjcJyb4BWCxGsndESCJQbdFMogUC5PbRDqceLTC
+&subject_token_type=
+  urn%3Aietf%3Aparams%3Aoauth%3Atoken-type%3Aaccess_token
+```
+
+Logto validates the subject token and other parameters in the token exchange request. If the request is valid, Logto generates a new access token that represents the identity of the user and returns it in the response.
+
+```
+HTTP/1.1 200 OK
+Content-Type: application/json
+Cache-Control: no-cache, no-store
+
+{
+  "access_token":"eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzd
+    WIiOiIxMjMiLCJzY29wZSI6Im9wZW5pZCBvZmZsaW5lX2FjY2VzcyBwc
+    m9maWxlIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyL
+    CJhY3QiOnsic3ViIjoiNDU2In19.H-X4W0mu-lJae27_rvsHAx5LpzsH
+    TDYozkWDWzBrMoM",
+  "scope":"openid offline_access profile",
+  "issued_token_type":
+    "urn:ietf:params:oauth:token-type:access_token",
+  "token_type":"Bearer",
+  "expires_in":3600
+}
+```
+
+The support engineer receives the access token and can now make API requests on behalf of the user.
+
+### Step 3: Investigate the issue
+
+The support engineer uses the access token to make API requests to investigate the issue reported by the user. The support engineer can verify the user's permissions, troubleshoot the problem, and provide assistance to the user.
```

Comment:
do we support organization token in this version?

Reply:
Yes, the token exchange flow will follow the same implementation of the normal flow.